enhancements/installer/aws-custom-region-and-endpoints.md

```diff
@@ -0,0 +1,402 @@
+  - "@enxebre"
+  - "@adambkaplan"
+  - "@dgoodwin"
+  - "@jstuever"
+  - "@patrickdillon"
+approvers:
+  - "@deads"
+  - "@sdodson"
+  - "@crawford"
+creation-date: 2019-12-13
+last-updated: 2020-03-03
+status: implementable
+---
+
+# AWS Custom Regions and Custom Endpoints Support
+
+## Release Signoff Checklist
+
+- [x]  Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+## Summary
+
+As an administrator, I would like to deploy OpenShift 4 clusters to AWS GovCloud, "hidden" regions and outpost deployments. These environments require the cluster to use custom endpoints for AWS APIs. The users provide the custom endpoints for various AWS services and the installer and cluster operators will use the specified endpoints to access the AWS APIs.
+
+## Motivation
+
+### Goals
+
+1. Allow users to specify API endpoints for some or all required AWS services.
+
+2. Provide users capability to install in custom regions on AWS.
+
+3. Provide automated handling for most users with respect to getting AMI available in the custom regions.
+
+### Non-Goals
+
+1. There are no plans to allow self-signed server certificates for endpoints specifically.
+
+2. Allowing users to use un-encrypted AMIs for the ec2 instances.
+
+## Proposal
+
+To support custom endpoints for AWS APIs, the install-config would allow the users to provide a list of endpoints for various services. When the custom endpoints are set, the installer will validate that the endpoints are reachable. The endpoints will be used by the installer to call AWS APIs for performing various actions like validations, creating the MachineSet objects, and also configuring the terraform AWS provider.
+
+When custom endpoints are provide for AWS APIs, the cluster operators also need the discover the information, therefore, the installer will make these available using the `config.openshift.io/v1` `Infrastructure` object. The cluster operators can then use the custom endpoints from the object to configure the AWS SDK to use the corresponding endpoints for communications.
+
+Since various kubernetes components like the kube-apiserver, kubelet (machine-config-operator), kube-controller-manager, cloud-controller-managers use the `.spec.cloudConfig` Config Map reference for cloud provider specific configurations, a new controller `cluster-kube-cloud-config-operator` will be introduced to perform the task of stitching the custom endpoints with the rest of the cloud config, such that all the kubernetes components can continue to directly consume a Config Map for configuration. This controller will perform the specialized stitching on the bootstrap host for control-plane kubelet and also actively reconcile the state in the cluster.
```

Comment:
to make sure I understand, would the `cluster-kube-cloud-config-operator` permanently watch changes to the custom endpoints and reconcile the cloud provider ConfigMap with them?